extensions/amp-analytics/0.1/analytics-root.js

```diff
@@ -0,0 +1,317 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {closestBySelector, matches} from '../../../src/dom';
+import {dev, user} from '../../../src/log';
+import {map} from '../../../src/utils/object';
+
+const TAG = 'amp-analytics';
+
+
+/**
+ * An analytics root. Analytics can be scoped to either ampdoc, embed or
+ * an arbitrary AMP element.
+ *
+ * TODO(dvoytenko): consider moving this concept into core as `AmpRoot`
+ * interface that will be implemented by `AmpDoc` and `FriendlyIframeEmbed`.
+ *
+ * @implements {../../../src/service.Disposable}
+ * @abstract
+ */
+export class AnalyticsRoot {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {?AnalyticsRoot} parent
+   */
+  constructor(ampdoc, parent) {
+    /** @const */
+    this.ampdoc = ampdoc;
+
+    /** @const */
+    this.parent = parent;
```

Comment:
It's needed for VisibilityEventTracker. I stripped it out of this PR though - too big. I can remove until we get there if you'd like.